apps/framework-cli/src/framework/core/state_storage.rs

```diff
@@ -0,0 +1,390 @@
+            .await
+            .context("Failed to create state table")?;
+
+        Ok(())
+    }
+
+    /// Try to acquire migration lock
+    /// Must be manually released with release_migration_lock()
+    /// Lock automatically expires after 5 minutes as a safety fallback
+    pub async fn acquire_migration_lock(&self) -> Result<()> {
+        self.ensure_state_table().await?;
+
+        // Enable strict mode for this session - INSERT will fail if key exists (not overwrite)
+        self.client
+            .client
+            .query("SET keeper_map_strict_mode = 1")
+            .execute()
+            .await
+            .context("Failed to enable strict mode")?;
+
+        // Check if lock exists
+        let existing_lock_query = format!(
+            "SELECT value FROM `{}`.`{}` WHERE key = '{}'",
+            self.db_name,
+            Self::STATE_TABLE,
+            Self::LOCK_KEY
+        );
+
+        let mut cursor = self
+            .client
+            .client
+            .query(&existing_lock_query)
+            .fetch::<String>()
+            .context("Failed to query for existing lock")?;
+
+        if let Ok(Some(lock_json_base64)) = cursor.next().await {
+            // Lock exists - check if expired
+            // Base64 decode the lock data
+            let lock_json_bytes = base64::Engine::decode(
+                &base64::engine::general_purpose::STANDARD,
+                lock_json_base64.as_bytes(),
+            )
+            .context("Failed to base64 decode lock data")?;
+            let lock_json = String::from_utf8(lock_json_bytes)
+                .context("Failed to convert lock data to UTF-8")?;
+
+            let existing_lock: MigrationLock =
+                serde_json::from_str(&lock_json).context("Failed to deserialize existing lock")?;
+
+            if existing_lock.expires_at < Utc::now() {
+                // Stale lock - delete it
+                let delete_sql = format!(
+                    "DELETE FROM `{}`.`{}` WHERE key = '{}'",
+                    self.db_name,
+                    Self::STATE_TABLE,
+                    Self::LOCK_KEY
+                );
+
+                self.client
+                    .client
+                    .query(&delete_sql)
+                    .execute()
+                    .await
+                    .context("Failed to delete stale lock")?;
+
+                warn!(
+                    "Deleted stale migration lock from machine {} (expired at {})",
+                    existing_lock.machine_id, existing_lock.expires_at
+                );
+            } else {
+                // Active lock held by someone else
+                let time_remaining = existing_lock.expires_at - Utc::now();
+                let minutes = time_remaining.num_minutes();
+                let seconds = time_remaining.num_seconds() % 60;
+
+                anyhow::bail!(
+                    "Migration already in progress on machine {}. Started at {}. Expires in {}m {}s.",
```

Comment:
In the docs we need a way for someone to manually delete a lock - it happens with terraform when the lock is fudged. You can manually unlock

Can be a follow up